### PR TITLE
use local image name for pull policy checks

### PR DESCRIPTION
**Pull request by vrothberg**
Some pull policies require to first look up a local image and compare
that to the remote counter part.  When looking up the remote image, we
need to make sure to use the name of the local image, if it exists.

This fixes a bug where a short name resolved to an image with the
"localhost/" prefix.  This prefix is only used for local image look ups
via `shortnames.ResolveLocally`.  Hence, when looking up the remote
counter part, we must preserve this prefix.

Fixes: #2904
Signed-off-by: Valentin Rothberg <rothberg@redhat.com>

### CODE REVIEW

**Comment by vrothberg** [comment]
@mtrmac @rhatdan @TomSweeneyRedHat PTAL

**Comment by rhatdan** [comment]
LGTM
Yes we need to get this into podman 3.0

**Review by giuseppe — APPROVED** [review]
/lgtm

File: new.go [hunk 1]
```diff
@@ -166,6 +166,12 @@ func resolveImage(ctx context.Context, systemContext *types.SystemContext, store
 		}
 	}
 
+	// If we found a local image, we must use it's name.
+	// See #2904.
+	if localImageRef != nil {
+		fromImage = localImageName
+	}
+
 	resolved, err := shortnames.Resolve(systemContext, fromImage)
```
Comment (mtrmac) [inline new.go:175]: This is now stuttering and hard to follow.

How about
```go
if options.PullPolicy == PullNever && localImage == nil { /* fail*/ }
if localImage != nil {
   // the three checks
}
```

File: new.go [hunk 2]
```diff
@@ -145,7 +145,7 @@ func resolveImage(ctx context.Context, systemContext *types.SystemContext, store
 		}
 	}
 
-	localImageRef, _, localImage, err := resolveLocalImage(systemContext, store, options)
+	localImageRef, _, localImageName, localImage, err := resolveLocalImage(systemContext, store, options)
```
Comment (mtrmac) [inline new.go:148]: Unrelated? This is the only caller, yet we ignore the second return value. It should be dropped, then.

File: new.go [hunk 3]
```diff
@@ -166,6 +166,12 @@ func resolveImage(ctx context.Context, systemContext *types.SystemContext, store
 		}
 	}
 
+	// If we found a local image, we must use it's name.
+	// See #2904.
+	if localImageRef != nil {
+		fromImage = localImageName
```
Comment (mtrmac) [inline new.go:172]: (Non-blocking: It might be useful to add a debug log if `fromImage != localImageName`, so that users can figure out where the fully-qualified variant came from.)

**Review by mtrmac — COMMENTED** [review]
Great point, with things like `PullIfNewer` we should always only use a remote image from the same registry as the local one.

File: new.go [hunk 4]
```diff
@@ -166,6 +166,12 @@ func resolveImage(ctx context.Context, systemContext *types.SystemContext, store
...
+	}
```
Comment (TomSweeneyRedHat) [inline new.go:173]: Neat workaround, hadn't thought about trying this.
